chain/rust/src/builders/tx_builder.rs

```diff
@@ -97,181 +183,141 @@ pub enum CoinSelectionStrategyCIP2 {
+        self.collateral_percentage = Some(collateral_percentage);
+        self
     }
 
-    pub fn collateral_percentage(&self, collateral_percentage: u32) -> Self {
-        let mut cfg = self.clone();
-        cfg.collateral_percentage = Some(collateral_percentage);
-        cfg
+    pub fn max_collateral_inputs(mut self, max_collateral_inputs: u32) -> Self {
+        self.max_collateral_inputs = Some(max_collateral_inputs);
+        self
     }
 
-    pub fn max_collateral_inputs(&self, max_collateral_inputs: u32) -> Self {
-        let mut cfg = self.clone();
-        cfg.max_collateral_inputs = Some(max_collateral_inputs);
-        cfg
-    }
-
-    pub fn build(&self) -> Result<TransactionBuilderConfig, JsError> {
-        let cfg = self.clone();
+    pub fn build(self) -> Result<TransactionBuilderConfig, TxBuilderError> {
         Ok(TransactionBuilderConfig {
-            fee_algo: cfg.fee_algo.ok_or_else(|| JsError::from_str("uninitialized field: fee_algo"))?,
-            pool_deposit: cfg.pool_deposit.ok_or_else(|| JsError::from_str("uninitialized field: pool_deposit"))?,
-            key_deposit: cfg.key_deposit.ok_or_else(|| JsError::from_str("uninitialized field: key_deposit"))?,
-            max_value_size: cfg.max_value_size.ok_or_else(|| JsError::from_str("uninitialized field: max_value_size"))?,
-            max_tx_size: cfg.max_tx_size.ok_or_else(|| JsError::from_str("uninitialized field: max_tx_size"))?,
-            coins_per_utxo_byte: cfg.coins_per_utxo_byte.ok_or_else(|| JsError::from_str(
-                "uninitialized field: coins_per_utxo_byte",
-            ))?,
-            /// TODO: remove once Babbage is on mainnet
-            coins_per_utxo_word: cfg.coins_per_utxo_word,
-            ex_unit_prices: cfg
+            fee_algo: self.fee_algo.ok_or_else(|| TxBuilderError::UninitializedField(TxBuilderConfigField::FeeAlgo))?,
+            pool_deposit: self.pool_deposit.ok_or_else(|| TxBuilderError::UninitializedField(TxBuilderConfigField::PoolDeposit))?,
+            key_deposit: self.key_deposit.ok_or_else(|| TxBuilderError::UninitializedField(TxBuilderConfigField::KeyDeposit))?,
+            max_value_size: self.max_value_size.ok_or_else(|| TxBuilderError::UninitializedField(TxBuilderConfigField::MaxValueSize))?,
+            max_tx_size: self.max_tx_size.ok_or_else(|| TxBuilderError::UninitializedField(TxBuilderConfigField::MaxTxSize))?,
+            coins_per_utxo_byte: self.coins_per_utxo_byte.ok_or_else(|| TxBuilderError::UninitializedField(TxBuilderConfigField::CoinsPerUtxoBytes))?,
+            ex_unit_prices: self
                 .ex_unit_prices
-                .ok_or_else(|| JsError::from_str("uninitialized field: ex_unit_prices"))?,
-            costmdls: if cfg.costmdls.is_some() {
-                cfg.costmdls.unwrap()
+                .ok_or_else(|| TxBuilderError::UninitializedField(TxBuilderConfigField::ExUnitPrices))?,
```

Comment:
why collateral and ex units are always required for tx builder config? not all the transactions need these params. besides that, do you know why we count fee with ex units only while balancing the change, why it is not done e.g. while selecting inputs?

Reply:
I just adapted how the code was from the old builders. I assume we just made it mandatory since you'd probably have all the protocol parameters available. We could probably refactor the builders to throw this error later on e.g. if you tried making a tx that used plutus stuff but didn't initialize it so you don't have to provide it always. I'm not sure about the ex units when selecting inputs. What would be needing it for input selection fee-wise that we could reasonably know? We treat plutus as a black box of bytes since we don't have access to an interpreter/etc. It shouldn't affect CBOR sizes since AFAIK the only place that uses it CBOR-wise would be as part of a hash so the size wouldn't change regardless of what the exunits are. I could be misunderstanding or overlooking something though. I just based these changes on the existing code we had in the old crate.